spec.md

```diff
@@ -8,121 +8,162 @@ title: Semantic Specification
     [Span C········································]
-         [Span E···········]   [Span F········]
-                                [Span G······]
-                                [Span H··]
-                                        [Span I·]
+         [Span E·······]        [Span F··] [Span G··] [Span H··]
 ~~~
 
 ## Traces
 
-A **Trace** represents the potentially distributed, potentially concurrent data/execution path in a (potentially distributed, potentially concurrent) system. A Trace can be thought of as a tree of Spans. (See the ASCII diagrams above)
+A **Trace** represents the potentially distributed, potentially concurrent data/execution path in a (potentially distributed, potentially concurrent) system. A Trace can be thought of as a directed acyclic graph (DAG) of Spans.
 
 ## Spans
 
-A **Span** represents a logical unit of work in the system that has a start time and a duration. Spans may be nested and ordered to model parent-child and causal relationships. Each span has an **operation name**, a presumably human-readable string which concisely names the work done by the span (e.g., an RPC method name, a function name, or the name of a subtask within a larger computation).
+A **Span** represents a logical unit of work in the system that has a start time and a duration. Spans may be nested and ordered to model causal relationships. Each span has an **operation name**, a presumably human-readable string which concisely represents the work done by the span (e.g., an RPC method name, a function name, or the name of a subtask or stage within a larger computation).
+
+<span id="references"></span>
+
+### Causal Span References
+
+A Span may reference zero or more Spans that are causally related. OpenTracing presently defines two types of references:
+
+**Parent references:** A "parent" Span may create a new "child" Span. The expectation is that the parent Span has not finished by the time the new Span starts. Though it is often the case that the parent Span cannot complete successfully until the child Span does as well, this is not a semantic requirement for parent references.
+
+A timing diagram for a parent that's blocked on a child Span:
+
+~~~
+    [-Parent Span---------]
+         [-Child Span----]
+~~~
+
+A timing diagram for a parent that's not blocked on a child Span:
+
+~~~
+    [-Parent Span-]
+              [-Child Span-]
+~~~
+
+**Sequence predecessor references:** Some Spans cannot start until a "sequence predecessor" Span finishes. For instance, chained sequence predecessor references neatly describe pipeline stages separated by queues.
+
+A sequence predecessor Span is part of the same logical trace as the descendant Span that refers to it: i.e., the descendant Span is formally caused by the work of its sequence predecessor.
+
+A timing diagram for a SequencePredecessorRef that must finish before the descendant Span can start:
+
+~~~
+    [-SequencePredecessorRef Span-]  [-New Span-]
+~~~
+
 
```

Comment:
Yeah this is an interesting use-case.  A single ID across all 4 stages has the advantage of grepability in logs, which is nice.

However, in our work we are leaning towards modeling this slightly differently.  One thing we found in trying to support cases like this is on the tracer collector/analysis side, it's much easier if the "single trace" that spans all 4 of those spans is actually 4 separate traces which refer to each other via a "parent trace" reference.  This helps us figure quickly out when "done" happens for each segment--a need that might be specific to the types of analysis we do.

The two approaches require about the same level of consideration on the instrumenter's part: the insight that the unit of work in each subsequent parent span is decoupled, and therefore the trace in the process should be started not as a standard continuation of an existing trace but as either a "sequence predecessor" reference or "parent trace" reference.

So these are very nearly 6 and 1/2 dozen (just different words) but the difference is whether the trace ID changes (or can change) at the juncture as well.  Curious what you think of the multi-ID model?